### DISCUSSION TITLE
Draggable handle wrapper

### DISCUSSION THREAD

**Discussion by wagnerpaz** [ueberdosis/tiptap#1811] [General Questions]
Hi! I'm trying to implement a draggable handle for every element in the editor (similar to Notion). For now I'm trying to extend Paragraph and Heading to add some wrapper divs that shows the handles.

I'm using React:

DraggableItemComponent:
```
import { NodeViewWrapper, NodeViewContent } from '@tiptap/react';

const DraggableItemComponent = () => {
  return (
    <NodeViewWrapper style={styles.draggableItem}>
      <div className="container">
        <div className="handle" style={styles.dragHandle} contentEditable="false" draggable="true" data-drag-handle />
        <div draggable="false">
          <NodeViewContent />
        </div>
        <style jsx>{`
          .handle {
            display: none;
          }
          .container {
            display: flex;
          }
          .container:hover .handle {
            display: block;
          }
        `}</style>
      </div>
    </NodeViewWrapper>
  );
};

const styles = {
  draggableItem: {
    display: 'flex',
    padding: '0.5rem',
    margin: '0.5rem 0',
    background: 'white',
    alignItems: 'center'
  },

  dragHandle: {
    flex: '0 0 auto',
    position: 'relative',
    width: '1rem',
    height: '1rem',
    top: '0.1rem',
    marginLeft: '-1.5rem',
    paddingRight: '0.5rem',
    cursor: 'grab',
    backgroundImage:
      'url(\'data:image/svg+xml;charset=UTF-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 10 16"><path fill-opacity="0.2" d="M4 14c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zM2 6C.9 6 0 6.9 0 8s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6C.9 0 0 .9 0 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z" /></svg>\')',
    backgroundRepeat: 'no-repeat',
    backgroundSize: 'contain',
    backgroundPosition: 'center'
  }
};

export default DraggableItemComponent;

```

DraggableHeading:

```
import Heading from '@tiptap/extension-heading';
import { ReactNodeViewRenderer } from '@tiptap/react';

import DraggableItemComponent from './DraggableItemComponent';

export default Heading.extend({
  draggable: true,

  addNodeView() {
    return ReactNodeViewRenderer(DraggableItemComponent);
  }
});

```

The problem is that the content inside de NodeView does not have the heading tags, just pure text. It shouldn't use the renderHTML of the Extension?

**Comment by philippkuehn** [comment]
Nodeviews overwrites the visual rendering in the browser of your node. Under the hood `renderHTML` is still needed and used for defining the schema, copy/pasting or exporting the HTML.

But you can set the tag of your nodeview with the `as` prop:

```jsx
<NodeViewWrapper as="h1">
 // ...
</NodeViewWrapper>
```

**Reply by kaspnilsson** [reply]
I'm also trying to solve this generically -- hoping to build a GutterMenu similar to FloatingMenu but that sticks to the LHS gutter as a community extension. There are tons of folks here posting about this in relation to global drag handles, persistent block add / delete buttons, etc... Basically most of what Notion has.

Found these related issues: 
- https://github.com/ueberdosis/tiptap/discussions/2095
- https://github.com/ueberdosis/tiptap/issues/1887
- https://github.com/ueberdosis/tiptap/issues/323

It doesn't seem correct or feasible to fetch the correct tag here from the node schema, especially when trying to do this for _all_ block elements and not just Paragraph / Heading.

Is there any way to call `renderHtml()` within my `ReactNodeViewRenderer`? Or, are there other, better recommendations for how we should go about building something like this?

**Comment by wagnerpaz** [comment]
I'm trying different approachs here to achieve my goal.
Another questions... how in renderHTML can I render nested tags? for example:
```
<div class="my-class"><p>
```

**Reply by philippkuehn** [reply]
You can see that at the codeblock extension: https://github.com/ueberdosis/tiptap/blob/main/packages/extension-code-block/src/code-block.ts

**Reply by hanspagel** [reply]
And in the docs:
https://www.tiptap.dev/guide/custom-extensions/#render-html
